File: 2022/day3/c++/part2.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

const int lower_alpha_ascii = 'a' - 1, upper_alpha_ascii = 'A' - 27;

int main() {
  ifstream input("../input.txt");

  if (!input.is_open()) {
    cerr << "Unable to open input file\n";

    return 1;
  }

  size_t priority_sum = 0;

  string item_list;
  vector<string> group;
  while (getline(input, item_list)) {
    group.push_back(item_list);

    if (group.size() != 3) continue;

    for (int i = 0, count = 1; i < 3; i++) {
      const string item_list_ = group[i];

      for (int j = 0; j < item_list_.length(); j++) {
        const char item = item_list_[j];

        for (int m = 0; m < 3; m++) {
          if (i == m) continue;
          if (group[m].find(item) == string::npos) break;

          count++;
        }

        if (count == 3) {
          priority_sum += item - (item < lower_alpha_ascii ? upper_alpha_ascii
                                                           : lower_alpha_ascii);
          break;
        }

        count = 1;
      }

      if (count == 3) break;
    }

    group.resize(0);
  }

  input.close();

  cout << priority_sum << '\n';
}

File: 2022/day3/javascript/part2.mjs
import { readFileSync } from 'node:fs';

const input = readFileSync('../input.txt', 'utf-8');

const itemLists = input.split(/\r?\n/);

let prioritySum = 0;

const lowerAlphaAscii = 'a'.charCodeAt(0) - 1;
const upperAlphaAscii = 'A'.charCodeAt(0) - 27;

for (let i = 0, group = []; i < itemLists.length; i++) {
  group.push(itemLists[i]);

  if (group.length !== 3) continue;

  for (let j = 0, count = 1; j < 3; j++) {
    const itemList = group[j];

    for (let m = 0; m < itemList.length; m++) {
      const item = itemList[m];

      for (let n = 0; n < 3; n++) {
        if (n === j) continue;
        if (!group[n].includes(item)) break;

        count++;
      }

      if (count === 3) {
        const asciiValue = item.charCodeAt(0);

        prioritySum +=
          asciiValue -
          (asciiValue < lowerAlphaAscii ? upperAlphaAscii : lowerAlphaAscii);
        break;
      }

      count = 1;
    }

    if (count === 3) break;
  }

  group = [];
}

console.log(prioritySum);
